File: packages/sdk-go/discord_gateway.go
package pocketping

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

// Discord Gateway opcodes
const (
	gatewayOpcodeDispatch        = 0
	gatewayOpcodeHeartbeat       = 1
	gatewayOpcodeIdentify        = 2
	gatewayOpcodeResume          = 6
	gatewayOpcodeReconnect       = 7
	gatewayOpcodeInvalidSession  = 9
	gatewayOpcodeHello           = 10
	gatewayOpcodeHeartbeatAck    = 11
)

// Discord Gateway intents
const (
	IntentGuilds                 = 1 << 0
	IntentGuildMessages          = 1 << 9
	IntentMessageContent         = 1 << 15
)

// DiscordGatewayConfig holds configuration for the Discord Gateway
type DiscordGatewayConfig struct {
	BotToken           string
	ChannelID          string // Forum channel ID for threads
	AllowedBotIDs      []string
	OnOperatorMessage  func(ctx context.Context, sessionID, content, operatorName string, attachments []Attachment, replyToBridgeMessageID *int)
	OnOperatorMessageWithIDs func(ctx context.Context, sessionID, content, operatorName string, attachments []Attachment, replyToBridgeMessageID *int, bridgeMessageID string)
	OnOperatorMessageEdit    func(ctx context.Context, sessionID, bridgeMessageID, content string, editedAt time.Time)
	OnOperatorMessageDelete  func(ctx context.Context, sessionID, bridgeMessageID string, deletedAt time.Time)
}

// DiscordGateway manages a persistent WebSocket connection to Discord Gateway
type DiscordGateway struct {
	config     DiscordGatewayConfig
	conn       *websocket.Conn
	sessionID  string
	resumeURL  string
	sequence   *int
	httpClient *http.Client

	heartbeatInterval time.Duration
	heartbeatTicker   *time.Ticker
	lastHeartbeatAck  time.Time

	mu        sync.Mutex
	connected bool
	ctx       context.Context
	cancel    context.CancelFunc
}

// gatewayPayload represents a Gateway payload
type gatewayPayload struct {
	Op int             `json:"op"`
	D  json.RawMessage `json:"d,omitempty"`
	S  *int            `json:"s,omitempty"`
	T  string          `json:"t,omitempty"`
}

// helloPayload represents the HELLO event data
type helloPayload struct {
	HeartbeatInterval int `json:"heartbeat_interval"`
}

// identifyPayload represents the IDENTIFY payload
type identifyPayload struct {
	Token      string            `json:"token"`
	Intents    int               `json:"intents"`
	Properties identifyProperties `json:"properties"`
}

type identifyProperties struct {
	OS      string `json:"os"`
	Browser string `json:"browser"`
	Device  string `json:"device"`
}

// readyPayload represents the READY event data
type readyPayload struct {
	SessionID string `json:"session_id"`
	ResumeGatewayURL string `json:"resume_gateway_url"`
}

// messageCreatePayload represents a MESSAGE_CREATE event
type messageCreatePayload struct {
	ID              string            `json:"id"`
	ChannelID       string            `json:"channel_id"`
	Content         string            `json:"content"`
	Author          discordUser       `json:"author"`
	Attachments     []discordAttachment `json:"attachments"`
	MessageReference *messageReference `json:"message_reference,omitempty"`
}

type messageUpdatePayload struct {
	ID             string       `json:"id"`
	ChannelID      string       `json:"channel_id"`
	Content        string       `json:"content,omitempty"`
	EditedTimestamp string      `json:"edited_timestamp,omitempty"`
	Author         *discordUser `json:"author,omitempty"`
}

type messageDeletePayload struct {
	ID        string `json:"id"`
	ChannelID string `json:"channel_id"`
	GuildID   string `json:"guild_id,omitempty"`
}

type discordUser struct {
	ID       string `json:"id"`
	Username string `json:"username"`
	Bot      bool   `json:"bot"`
}

type discordAttachment struct {
	ID          string `json:"id"`
	Filename    string `json:"filename"`
	ContentType string `json:"content_type"`
	Size        int    `json:"size"`
	URL         string `json:"url"`
}

type messageReference struct {
	MessageID string `json:"message_id,omitempty"`
	ChannelID string `json:"channel_id,omitempty"`
	GuildID   string `json:"guild_id,omitempty"`
}

// NewDiscordGateway creates a new Discord Gateway instance
func NewDiscordGateway(config DiscordGatewayConfig) *DiscordGateway {
	return &DiscordGateway{
		config:     config,
		httpClient: &http.Client{Timeout: 30 * time.Second},
	}
}

// Connect establishes a connection to the Discord Gateway
func (g *DiscordGateway) Connect(ctx context.Context) error {
	g.ctx, g.cancel = context.WithCancel(ctx)

	// Get Gateway URL
	gatewayURL, err := g.getGatewayURL()
	if err != nil {
		return fmt.Errorf("get gateway URL: %w", err)
	}

	// Connect to WebSocket
	conn, _, err := websocket.DefaultDialer.DialContext(g.ctx, gatewayURL+"?v=10&encoding=json", nil)
	if err != nil {
		return fmt.Errorf("connect to gateway: %w", err)
	}
	g.conn = conn
	g.connected = true

	// Start listening for messages
	go g.listen()

	return nil
}

// Close closes the Gateway connection
func (g *DiscordGateway) Close() error {
	g.mu.Lock()
	defer g.mu.Unlock()

	if g.cancel != nil {
		g.cancel()
	}

	if g.heartbeatTicker != nil {
		g.heartbeatTicker.Stop()
	}

	if g.conn != nil {
		g.connected = false
		return g.conn.Close()
	}

	return nil
}

func (g *DiscordGateway) getGatewayURL() (string, error) {
	if g.resumeURL != "" {
		return g.resumeURL, nil
	}

	req, err := http.NewRequestWithContext(g.ctx, "GET", "https://discord.com/api/v10/gateway/bot", nil)
	if err != nil {
		return "", err
	}
	req.Header.Set("Authorization", "Bot "+g.config.BotToken)

	resp, err := g.httpClient.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	var result struct {
		URL string `json:"url"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return "", err
	}

	return result.URL, nil
}

func (g *DiscordGateway) listen() {
	for {
		select {
		case <-g.ctx.Done():
			return
		default:
		}

		_, message, err := g.conn.ReadMessage()
		if err != nil {
			if websocket.IsCloseError(err, websocket.CloseNormalClosure, websocket.CloseGoingAway) {
				return
			}
			log.Printf("[DiscordGateway] Read error: %v", err)
			g.reconnect()
			return
		}

		var payload gatewayPayload
		if err := json.Unmarshal(message, &payload); err != nil {
			log.Printf("[DiscordGateway] Parse error: %v", err)
			continue
		}

		// Update sequence number
		if payload.S != nil {
			g.sequence = payload.S
		}

		g.handlePayload(payload)
	}
}

func (g *DiscordGateway) handlePayload(payload gatewayPayload) {
	switch payload.Op {
	case gatewayOpcodeHello:
		var hello helloPayload
		if err := json.Unmarshal(payload.D, &hello); err != nil {
			log.Printf("[DiscordGateway] Parse HELLO error: %v", err)
			return
		}
		g.heartbeatInterval = time.Duration(hello.HeartbeatInterval) * time.Millisecond
		g.startHeartbeat()
		g.identify()

	case gatewayOpcodeHeartbeatAck:
		g.lastHeartbeatAck = time.Now()

	case gatewayOpcodeHeartbeat:
		g.sendHeartbeat()

	case gatewayOpcodeReconnect:
		g.reconnect()

	case gatewayOpcodeInvalidSession:
		// Wait and re-identify
		time.Sleep(5 * time.Second)
		g.sessionID = ""
		g.sequence = nil
		g.identify()

	case gatewayOpcodeDispatch:
		g.handleDispatch(payload.T, payload.D)
	}
}

func (g *DiscordGateway) handleDispatch(eventType string, data json.RawMessage) {
	switch eventType {
	case "READY":
		var ready readyPayload
		if err := json.Unmarshal(data, &ready); err != nil {
			log.Printf("[DiscordGateway] Parse READY error: %v", err)
			return
		}
		g.sessionID = ready.SessionID
		g.resumeURL = ready.ResumeGatewayURL
		log.Printf("[DiscordGateway] Connected, session: %s", g.sessionID)

	case "MESSAGE_CREATE":
		var msg messageCreatePayload
		if err := json.Unmarshal(data, &msg); err != nil {
			log.Printf("[DiscordGateway] Parse MESSAGE_CREATE error: %v", err)
			return
		}
		g.handleMessage(msg)
	case "MESSAGE_UPDATE":
		var msg messageUpdatePayload
		if err := json.Unmarshal(data, &msg); err != nil {
			log.Printf("[DiscordGateway] Parse MESSAGE_UPDATE error: %v", err)
			return
		}
		g.handleMessageUpdate(msg)
	case "MESSAGE_DELETE":
		var msg messageDeletePayload
		if err := json.Unmarshal(data, &msg); err != nil {
			log.Printf("[DiscordGateway] Parse MESSAGE_DELETE error: %v", err)
			return
		}
		g.handleMessageDelete(msg)
	}
}

func (g *DiscordGateway) handleMessage(msg messageCreatePayload) {
	// Skip bot messages
	if msg.Author.Bot && !g.isAllowedBot(msg.Author.ID) {
		return
	}

	// Only process messages in threads under our channel
	// The channelID in the message is the thread ID
	// We should check if the parent is our forum channel
	// For now, we process all non-bot messages

	if g.config.OnOperatorMessage == nil && g.config.OnOperatorMessageWithIDs == nil {
		return
	}

	// Download attachments
	var attachments []Attachment
	for _, att := range msg.Attachments {
		data, err := g.downloadFile(att.URL)
		if err != nil {
			log.Printf("[DiscordGateway] Failed to download attachment: %v", err)
			continue
		}
		attachments = append(attachments, Attachment{
			Filename: att.Filename,
			MimeType: att.ContentType,
			Size:     int64(att.Size),
			Data:     data,
		})
	}

	// Get reply reference if present
	var replyToBridgeMessageID *int
	// Discord uses string IDs, but we need to convert for consistency
	// For Discord, we'll pass nil since the ID is a string snowflake, not int
	// The backend will need to handle this differently

	// Call the callback
	if g.config.OnOperatorMessage != nil {
		g.config.OnOperatorMessage(
			context.Background(),
			msg.ChannelID, // Thread/channel ID as session ID
			msg.Content,
			msg.Author.Username,
			attachments,
			replyToBridgeMessageID,
		)
	}
	if g.config.OnOperatorMessageWithIDs != nil {
		g.config.OnOperatorMessageWithIDs(
			context.Background(),
			msg.ChannelID,
			msg.Content,
			msg.Author.Username,
			attachments,
			replyToBridgeMessageID,
			msg.ID,
		)
	}
}

func (g *DiscordGateway) handleMessageUpdate(msg messageUpdatePayload) {
	if msg.Author != nil && msg.Author.Bot && !g.isAllowedBot(msg.Author.ID) {
		return
	}

	if g.config.OnOperatorMessageEdit == nil {
		return
	}

	if msg.Content == "" {
		return
	}

	editedAt := time.Now()
	if msg.EditedTimestamp != "" {
		if parsed, err := time.Parse(time.RFC3339, msg.EditedTimestamp); err == nil {
			editedAt = parsed
		}
	}

	g.config.OnOperatorMessageEdit(context.Background(), msg.ChannelID, msg.ID, msg.Content, editedAt)
}

func (g *DiscordGateway) handleMessageDelete(msg messageDeletePayload) {
	if g.config.OnOperatorMessageDelete == nil {
		return
	}

	g.config.OnOperatorMessageDelete(context.Background(), msg.ChannelID, msg.ID, time.Now())
}

func (g *DiscordGateway) isAllowedBot(botID string) bool {
	if botID == "" {
		return false
	}
	for _, allowed := range g.config.AllowedBotIDs {
		if allowed == botID {
			return true
		}
	}
	return false
}

func (g *DiscordGateway) downloadFile(url string) ([]byte, error) {
	req, err := http.NewRequestWithContext(g.ctx, "GET", url, nil)
	if err != nil {
		return nil, err
	}

	resp, err := g.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("download failed with status %d", resp.StatusCode)
	}

	var data []byte
	buf := make([]byte, 1024)
	for {
		n, err := resp.Body.Read(buf)
		if n > 0 {
			data = append(data, buf[:n]...)
		}
		if err != nil {
			break
		}
	}

	return data, nil
}

func (g *DiscordGateway) identify() {
	identify := identifyPayload{
		Token:   g.config.BotToken,
		Intents: IntentGuilds | IntentGuildMessages | IntentMessageContent,
		Properties: identifyProperties{
			OS:      "linux",
			Browser: "pocketping",
			Device:  "pocketping",
		},
	}

	data, _ := json.Marshal(identify)
	payload := gatewayPayload{
		Op: gatewayOpcodeIdentify,
		D:  data,
	}

	g.send(payload)
}

func (g *DiscordGateway) startHeartbeat() {
	if g.heartbeatTicker != nil {
		g.heartbeatTicker.Stop()
	}

	g.heartbeatTicker = time.NewTicker(g.heartbeatInterval)
	g.lastHeartbeatAck = time.Now()

	go func() {
		for {
			select {
			case <-g.ctx.Done():
				return
			case <-g.heartbeatTicker.C:
				// Check if we received a heartbeat ACK
				if time.Since(g.lastHeartbeatAck) > g.heartbeatInterval*2 {
					log.Printf("[DiscordGateway] Heartbeat timeout, reconnecting...")
					g.reconnect()
					return
				}
				g.sendHeartbeat()
			}
		}
	}()
}

func (g *DiscordGateway) sendHeartbeat() {
	var data json.RawMessage
	if g.sequence != nil {
		data, _ = json.Marshal(*g.sequence)
	} else {
		data = json.RawMessage("null")
	}

	payload := gatewayPayload{
		Op: gatewayOpcodeHeartbeat,
		D:  data,
	}

	g.send(payload)
}

func (g *DiscordGateway) send(payload gatewayPayload) error {
	g.mu.Lock()
	defer g.mu.Unlock()

	if g.conn == nil {
		return fmt.Errorf("not connected")
	}

	data, err := json.Marshal(payload)
	if err != nil {
		return err
	}

	return g.conn.WriteMessage(websocket.TextMessage, data)
}

func (g *DiscordGateway) reconnect() {
	g.mu.Lock()
	if g.conn != nil {
		g.conn.Close()
		g.conn = nil
	}
	g.connected = false
	g.mu.Unlock()

	// Wait before reconnecting
	time.Sleep(5 * time.Second)

	// Try to reconnect
	for i := 0; i < 5; i++ {
		select {
		case <-g.ctx.Done():
			return
		default:
		}

		gatewayURL := g.resumeURL
		if gatewayURL == "" {
			var err error
			gatewayURL, err = g.getGatewayURL()
			if err != nil {
				log.Printf("[DiscordGateway] Failed to get gateway URL: %v", err)
				time.Sleep(time.Duration(i+1) * 5 * time.Second)
				continue
			}
		}

		conn, _, err := websocket.DefaultDialer.DialContext(g.ctx, gatewayURL+"?v=10&encoding=json", nil)
		if err != nil {
			log.Printf("[DiscordGateway] Reconnect attempt %d failed: %v", i+1, err)
			time.Sleep(time.Duration(i+1) * 5 * time.Second)
			continue
		}

		g.mu.Lock()
		g.conn = conn
		g.connected = true
		g.mu.Unlock()

		// Start listening again
		go g.listen()
		return
	}

	log.Printf("[DiscordGateway] Failed to reconnect after 5 attempts")
}

File: community/lib/bridges/discord-gateway.ts
/**
 * Discord Gateway - Persistent WebSocket connection for receiving messages
 * Community Edition
 *
 * This enables real-time message receiving from Discord operators.
 * Only works on persistent hosting (Docker, VPS) - NOT serverless.
 */

import WebSocket from 'ws'

const NS = 'DiscordGateway'
const DISCORD_GATEWAY_URL = 'wss://gateway.discord.gg/?v=10&encoding=json'

// Gateway opcodes
const OP_DISPATCH = 0
const OP_HEARTBEAT = 1
const OP_IDENTIFY = 2
const OP_RESUME = 6
const OP_RECONNECT = 7
const OP_INVALID_SESSION = 9
const OP_HELLO = 10
const OP_HEARTBEAT_ACK = 11

// Intents
const INTENT_GUILDS = 1 << 0
const INTENT_GUILD_MESSAGES = 1 << 9
const INTENT_MESSAGE_CONTENT = 1 << 15

export interface DiscordGatewayConfig {
  botToken: string
  channelId: string // Forum channel ID for threads
  allowedBotIds?: string[]
  onOperatorMessage: (params: {
    threadId: string
    content: string
    operatorName: string
    attachments: DiscordGatewayAttachment[]
    messageId: string
    replyToMessageId?: string
  }) => void | Promise<void>
  onOperatorMessageEdit?: (params: {
    threadId: string
    messageId: string
    content: string
    editedAt?: Date
  }) => void | Promise<void>
  onOperatorMessageDelete?: (params: {
    threadId: string
    messageId: string
    deletedAt?: Date
  }) => void | Promise<void>
}

export interface DiscordGatewayAttachment {
  id: string
  filename: string
  contentType: string
  size: number
  url: string
  proxyUrl: string
}

interface GatewayPayload {
  op: number
  d?: unknown
  s?: number | null
  t?: string | null
}

interface HelloPayload {
  heartbeat_interval: number
}

interface ReadyPayload {
  session_id: string
  resume_gateway_url: string
}

interface MessageCreatePayload {
  id: string
  channel_id: string
  content: string
  author: {
    id: string
    username: string
    bot?: boolean
  }
  attachments?: Array<{
    id: string
    filename: string
    content_type?: string
    size: number
    url: string
    proxy_url: string
  }>
  message_reference?: {
    message_id?: string
    channel_id?: string
    guild_id?: string
  }
}

interface MessageUpdatePayload {
  id: string
  channel_id: string
  content?: string
  edited_timestamp?: string
  author?: {
    id: string
    username: string
    bot?: boolean
  }
}

interface MessageDeletePayload {
  id: string
  channel_id: string
  guild_id?: string
}

export class DiscordGateway {
  private config: DiscordGatewayConfig
  private ws: WebSocket | null = null
  private sessionId: string | null = null
  private resumeUrl: string | null = null
  private sequence: number | null = null
  private heartbeatInterval: number | null = null
  private heartbeatTimer: NodeJS.Timeout | null = null
  private lastHeartbeatAck: boolean = true
  private reconnectAttempts = 0
  private maxReconnectAttempts = 5
  private isConnecting = false
  private isClosing = false

  constructor(config: DiscordGatewayConfig) {
    this.config = config
  }

  async connect(): Promise<void> {
    if (this.isConnecting || this.ws?.readyState === WebSocket.OPEN) {
      console.log('⚠️ [DiscordGateway] Already connected or connecting')
      return
    }

    this.isConnecting = true
    this.isClosing = false

    try {
      const url = this.resumeUrl || DISCORD_GATEWAY_URL
      console.log(`🔌 [DiscordGateway] Connecting to Discord Gateway...`)

      this.ws = new WebSocket(url)

      this.ws.on('open', () => {
        console.log('🟢 [DiscordGateway] WebSocket connected')
        this.reconnectAttempts = 0
      })

      this.ws.on('message', (data: WebSocket.Data) => {
        try {
          const payload = JSON.parse(data.toString()) as GatewayPayload
          this.handlePayload(payload)
        } catch (err) {
          console.error('🔴 [DiscordGateway] Failed to parse message:', err)
        }
      })

      this.ws.on('close', (code: number, reason: Buffer) => {
        console.log(`🔶 [DiscordGateway] WebSocket closed: ${code} - ${reason.toString()}`)
        this.stopHeartbeat()
        this.ws = null
        this.isConnecting = false

        if (!this.isClosing) {
          this.attemptReconnect()
        }
      })

      this.ws.on('error', (err: Error) => {
        console.error('🔴 [DiscordGateway] WebSocket error:', err.message)
      })
    } catch (err) {
      console.error('🔴 [DiscordGateway] Connection failed:', err)
      this.isConnecting = false
      this.attemptReconnect()
    }
  }

  close(): void {
    console.log('🔷 [DiscordGateway] Closing gateway connection')
    this.isClosing = true
    this.stopHeartbeat()

    if (this.ws) {
      this.ws.close(1000, 'Normal closure')
      this.ws = null
    }

    this.sessionId = null
    this.resumeUrl = null
    this.sequence = null
  }

  private handlePayload(payload: GatewayPayload): void {
    if (payload.s !== null && payload.s !== undefined) {
      this.sequence = payload.s
    }

    switch (payload.op) {
      case OP_HELLO:
        this.handleHello(payload.d as HelloPayload)
        break

      case OP_HEARTBEAT:
        this.sendHeartbeat()
        break

      case OP_HEARTBEAT_ACK:
        this.lastHeartbeatAck = true
        break

      case OP_RECONNECT:
        console.log('🔷 [DiscordGateway] Received reconnect request')
        this.reconnect()
        break

      case OP_INVALID_SESSION:
        const canResume = payload.d as boolean
        console.log(`🔷 [DiscordGateway] Invalid session, can resume: ${canResume}`)
        if (!canResume) {
          this.sessionId = null
          this.resumeUrl = null
          this.sequence = null
        }
        setTimeout(() => this.identify(), 5000)
        break

      case OP_DISPATCH:
        this.handleDispatch(payload.t!, payload.d)
        break
    }
  }

  private handleHello(data: HelloPayload): void {
    console.log(`🔷 [DiscordGateway] Hello received, heartbeat interval: ${data.heartbeat_interval}ms`)
    this.heartbeatInterval = data.heartbeat_interval
    this.startHeartbeat()

    if (this.sessionId && this.sequence !== null) {
      this.resume()
    } else {
      this.identify()
    }
  }

  private handleDispatch(eventType: string, data: unknown): void {
    switch (eventType) {
      case 'READY':
        const ready = data as ReadyPayload
        this.sessionId = ready.session_id
        this.resumeUrl = ready.resume_gateway_url
        this.isConnecting = false
        console.log(`✅ [DiscordGateway] Connected with session: ${this.sessionId}`)
        break

      case 'RESUMED':
        this.isConnecting = false
        console.log('✅ [DiscordGateway] Session resumed')
        break

      case 'MESSAGE_CREATE':
        this.handleMessageCreate(data as MessageCreatePayload)
        break
      case 'MESSAGE_UPDATE':
        this.handleMessageUpdate(data as MessageUpdatePayload)
        break
      case 'MESSAGE_DELETE':
        this.handleMessageDelete(data as MessageDeletePayload)
        break
    }
  }

  private handleMessageCreate(msg: MessageCreatePayload): void {
    // Skip bot messages
    if (msg.author.bot && !this.isAllowedBot(msg.author.id)) {
      console.log(`⏭️ [DiscordGateway] Skipping bot message from ${msg.author.username}`)
      return
    }

    console.log(`📩 [DiscordGateway] Message received in channel ${msg.channel_id} from ${msg.author.username}: "${msg.content.substring(0, 50)}..."`)

    // Convert attachments
    const attachments: DiscordGatewayAttachment[] = (msg.attachments || []).map((att) => ({
      id: att.id,
      filename: att.filename,
      contentType: att.content_type || 'application/octet-stream',
      size: att.size,
      url: att.url,
      proxyUrl: att.proxy_url,
    }))

    // Get reply reference if present
    const replyToMessageId = msg.message_reference?.message_id

    // Call the callback
    try {
      this.config.onOperatorMessage({
        threadId: msg.channel_id,
        content: msg.content,
        operatorName: msg.author.username,
        attachments,
        messageId: msg.id,
        replyToMessageId,
      })
    } catch (err) {
      console.error('🔴 [DiscordGateway] Error in onOperatorMessage callback:', err)
    }
  }

  private handleMessageUpdate(msg: MessageUpdatePayload): void {
    if (msg.author?.bot && !this.isAllowedBot(msg.author.id)) {
      return
    }

    if (!this.config.onOperatorMessageEdit) {
      return
    }

    if (msg.content === undefined) {
      return
    }

    try {
      this.config.onOperatorMessageEdit({
        threadId: msg.channel_id,
        messageId: msg.id,
        content: msg.content ?? '',
        editedAt: msg.edited_timestamp ? new Date(msg.edited_timestamp) : undefined,
      })
    } catch (err) {
      console.error('🔴 [DiscordGateway] Error in onOperatorMessageEdit callback:', err)
    }
  }

  private handleMessageDelete(msg: MessageDeletePayload): void {
    if (!this.config.onOperatorMessageDelete) {
      return
    }

    try {
      this.config.onOperatorMessageDelete({
        threadId: msg.channel_id,
        messageId: msg.id,
        deletedAt: new Date(),
      })
    } catch (err) {
      console.error('🔴 [DiscordGateway] Error in onOperatorMessageDelete callback:', err)
    }
  }

  private isAllowedBot(botId?: string): boolean {
    if (!botId) return false
    return this.config.allowedBotIds?.includes(botId) ?? false
  }

  private identify(): void {
    console.log('🔷 [DiscordGateway] Sending IDENTIFY')
    this.send({
      op: OP_IDENTIFY,
      d: {
        token: this.config.botToken,
        intents: INTENT_GUILDS | INTENT_GUILD_MESSAGES | INTENT_MESSAGE_CONTENT,
        properties: {
          os: 'linux',
          browser: 'pocketping-community',
          device: 'pocketping-community',
        },
      },
    })
  }

  private resume(): void {
    console.log('🔷 [DiscordGateway] Sending RESUME')
    this.send({
      op: OP_RESUME,
      d: {
        token: this.config.botToken,
        session_id: this.sessionId,
        seq: this.sequence,
      },
    })
  }

  private startHeartbeat(): void {
    this.stopHeartbeat()

    if (!this.heartbeatInterval) return

    this.lastHeartbeatAck = true
    this.heartbeatTimer = setInterval(() => {
      if (!this.lastHeartbeatAck) {
        console.log('⚠️ [DiscordGateway] Heartbeat timeout, reconnecting...')
        this.reconnect()
        return
      }

      this.lastHeartbeatAck = false
      this.sendHeartbeat()
    }, this.heartbeatInterval)
  }

  private stopHeartbeat(): void {
    if (this.heartbeatTimer) {
      clearInterval(this.heartbeatTimer)
      this.heartbeatTimer = null
    }
  }

  private sendHeartbeat(): void {
    this.send({
      op: OP_HEARTBEAT,
      d: this.sequence,
    })
  }

  private send(payload: GatewayPayload): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      console.log('⚠️ [DiscordGateway] Cannot send, WebSocket not open')
      return
    }

    this.ws.send(JSON.stringify(payload))
  }

  private reconnect(): void {
    this.stopHeartbeat()

    if (this.ws) {
      this.ws.close(4000, 'Reconnecting')
      this.ws = null
    }

    setTimeout(() => this.connect(), 5000)
  }

  private attemptReconnect(): void {
    if (this.isClosing || this.reconnectAttempts >= this.maxReconnectAttempts) {
      console.error(`🔴 [DiscordGateway] Max reconnect attempts (${this.maxReconnectAttempts}) reached`)
      return
    }

    this.reconnectAttempts++
    const delay = Math.min(1000 * Math.pow(2, this.reconnectAttempts), 30000)
    console.log(`🔷 [DiscordGateway] Reconnecting in ${delay}ms (attempt ${this.reconnectAttempts})`)

    setTimeout(() => this.connect(), delay)
  }
}

// Global gateway manager
const activeGateways = new Map<string, DiscordGateway>()

export function startGateway(projectId: string, config: DiscordGatewayConfig): DiscordGateway {
  const existing = activeGateways.get(projectId)
  if (existing) {
    existing.close()
  }

  const gateway = new DiscordGateway(config)
  activeGateways.set(projectId, gateway)
  gateway.connect()

  return gateway
}

export function stopGateway(projectId: string): void {
  const gateway = activeGateways.get(projectId)
  if (gateway) {
    gateway.close()
    activeGateways.delete(projectId)
  }
}

export function getGateway(projectId: string): DiscordGateway | undefined {
  return activeGateways.get(projectId)
}

export function stopAllGateways(): void {
  for (const gateway of activeGateways.values()) {
    gateway.close()
  }
  activeGateways.clear()
}
